src/hotspot/share/utilities/ostream.hpp

```diff
@@ -193,27 +193,31 @@ class ttyUnlocker: StackObj {
 // Buffer will always be zero-terminated.
 class stringStream : public outputStream {
  protected:
-  char*  buffer;
-  size_t buffer_pos;
-  size_t buffer_length;
-  bool   buffer_fixed;
+  char*  _buffer;
+  size_t _written;
+  size_t _capacity;
+  const bool _is_fixed;
+  char   _small_buffer[32];
```

Comment:
I'm a little surprised that 32 bytes is that effective. Any idea how much better
(in malloc avoidance) a few more words would be?  Since 32 bytes isn't really
all that much.

Reply:
I was surprised too. I did not want to blow up the stringStream object too much because I was afraid someone may use it on stack in a recursion. I even considered melting the buffer and the buffer pointer into a union to save space, since you only need either one or the other, but the coding got too complex and it felt overengineered.

I did some tests:

java -version

```
debug:    <32:1038 (86%), <48:111 (9%), <64:42 (3%), >=64:4 (0%)
release:  <32:449 (84%),  <48:54 (10%), <64:26 (4%), >=64:4 (0%)
```

Running some unrelated test program, no tracing, doing metaspace stuff

```
debug: 	  <32:151985 (52%), <48:83341 (28%), <64:46916 (16%), >=64:6445 (2%)                  
release:  <32:84123 (49%),  <48:80332 (47%), <64:871 (0%), >=64:3567 (2%) 
```

same test with -XX:+PrintOptoInlining, which creates many temporary stringStream objects with large buffers

```
debug:	  <32:160311 (52%), <48:83661 (27%), <64:47821 (15%), >=64:15345 (4%)   
```

So, depending on the scenario, increasing the buffer size to 48 may give us between 10% and almost 30% more cases. Beyond that the effect strongly diminishes. I will increase it to 48.